File: strings/vowel-spellchecker/vowel-spellchecker.js
/**
 * @param {string[]} wordlist
 * @param {string[]} queries
 * @return {string[]}
 */
function spellchecker(wordlist, queries) {
  const words = new Set(wordlist);
  const lowered = {};
  const noVowels = {};

  for (let i = 0; i < wordlist.length; i++) {
    const word = wordlist[i];
    const l = caseInsensitive(word);
    const e = errorVowels(word);

    if (!lowered[l]) {
      lowered[l] = word;
    }

    if (!noVowels[e]) {
      noVowels[e] = word;
    }
  }

  for (let i = 0; i < queries.length; i++) {
    const word = queries[i];

    if (words.has(word)) {
      continue;
    }

    const lower = caseInsensitive(word);
    const noVowel = errorVowels(word);

    if (lowered[lower]) {
      queries[i] = lowered[lower];
    } else if (noVowels[noVowel]) {
      queries[i] = noVowels[noVowel];
    } else {
      queries[i] = '';
    }
  }
  return queries;
}

/**
 *
 * @param {string} word
 * @returns {string}
 */
function caseInsensitive(word) {
  return word.toLowerCase();
}

/**
 *
 * @param {string} word
 * @returns {string}
 */
function errorVowels(word) {
  return word.toLowerCase().replace(/[aeiou]/g, '*');
}

File: strings/vowel-spellchecker/vowel-spellchecker.java
class Solution {
  public String[] spellchecker(String[] wordlist, String[] queries) {
    Set<String> words = new HashSet<>(Arrays.asList(wordlist));
    Map<String, String> lowered = new HashMap<>();
    Map<String, String> noVowels = new HashMap<>();

    for (String word : wordlist) {
      lowered.putIfAbsent(caseInsensitive(word), word);
      noVowels.putIfAbsent(errorVowels(word), word);
    }

    for (int i = 0; i < queries.length; i++) {
      String word = queries[i];

      if (words.contains(word)) {
        continue;
      }

      String lower = caseInsensitive(word);
      String noVowel = errorVowels(word);

      if (lowered.containsKey(lower)) {
        queries[i] = lowered.get(lower);
      } else if (noVowels.containsKey(noVowel)) {
        queries[i] = noVowels.get(noVowel);
      } else {
        queries[i] = "";
      }
    }
    return queries;
  }

  private String caseInsensitive(String word) {
    return word.toLowerCase();
  }

  private String errorVowels(String word) {
    return word.toLowerCase().replaceAll("[aeiou]", "*");
  }
}
